src/libostree/ostree-sign-x509.c

```diff
@@ -0,0 +1,651 @@
+}
+
+static void
+_ostree_sign_x509_init (OstreeSignX509 *self)
+{
+
+  self->state = X509_OK;
+  self->secret_key = NULL;
+  self->public_keys = NULL;
+  self->revoked_keys = NULL;
+
+#if !defined(USE_OPENSSL)
+  self->state = X509_NOT_SUPPORTED;
+#else
+  if (!otcore_x509_init ())
+    self->state = X509_FAILED_INITIALIZATION;
+#endif
+}
+
+static gboolean
+_ostree_sign_x509_is_initialized (OstreeSignX509 *self, GError **error)
+{
+  switch (self->state)
+    {
+    case X509_OK:
+      break;
+    case X509_NOT_SUPPORTED:
+      return glnx_throw (error, "x509: engine is not supported");
+    case X509_FAILED_INITIALIZATION:
+      return glnx_throw (error, "x509: crypto library isn't initialized properly");
+    }
+
+  return TRUE;
+}
+
+gboolean
+ostree_sign_x509_data (OstreeSign *self, GBytes *data, GBytes **signature,
+                       GCancellable *cancellable, GError **error)
+{
+
+  g_assert (OSTREE_IS_SIGN (self));
+  OstreeSignX509 *sign = _ostree_sign_x509_get_instance_private (OSTREE_SIGN_X509 (self));
+
+  if (!_ostree_sign_x509_is_initialized (sign, error))
+    return FALSE;
+
+  if (sign->secret_key == NULL)
+    return glnx_throw (error, "Not able to sign: secret key is not set");
+
+#if defined(USE_OPENSSL)
+  gsize secret_key_size;
+  const guint8 *secret_key_buf = g_bytes_get_data (sign->secret_key, &secret_key_size);
+#endif
+
+  unsigned long long sig_size = 0;
+  g_autofree guchar *sig = NULL;
+
+#if defined(USE_OPENSSL)
+  EVP_MD_CTX *ctx = EVP_MD_CTX_new ();
+  if (!ctx)
+    return glnx_throw (error, "openssl: failed to allocate context");
+
+  const unsigned char *p = secret_key_buf;
+  EVP_PKEY *pkey = d2i_AutoPrivateKey (NULL, &p, secret_key_size);
+  if (!pkey)
+    {
+      EVP_MD_CTX_free (ctx);
+      return glnx_throw (error, "openssl: Failed to initialize x509 key");
+    }
+
+  size_t len;
+  if (EVP_DigestSignInit (ctx, NULL, NULL, NULL, pkey)
+      && EVP_DigestSign (ctx, NULL, &len, g_bytes_get_data (data, NULL), g_bytes_get_size (data)))
+    {
+      sig = g_malloc0 (len);
+      if (EVP_DigestSign (ctx, sig, &len, g_bytes_get_data (data, NULL), g_bytes_get_size (data)))
+        sig_size = len;
+    }
+
+  EVP_PKEY_free (pkey);
+  EVP_MD_CTX_free (ctx);
+
+#endif
+
+  if (sig_size == 0)
+    return glnx_throw (error, "Failed to sign");
+
+  *signature = g_bytes_new_take (g_steal_pointer (&sig), sig_size);
+  return TRUE;
+}
+
+gboolean
+ostree_sign_x509_data_verify (OstreeSign *self, GBytes *data, GVariant *signatures,
+                              char **out_success_message, GError **error)
+{
+  g_assert (OSTREE_IS_SIGN (self));
+
+  if (data == NULL)
+    return glnx_throw (error, "x509: unable to verify NULL data");
+
+  OstreeSignX509 *sign = _ostree_sign_x509_get_instance_private (OSTREE_SIGN_X509 (self));
+
+  if (!_ostree_sign_x509_is_initialized (sign, error))
+    return FALSE;
+
+  if (signatures == NULL)
+    return glnx_throw (error, "x509: commit have no signatures of my type");
```

Comment:
Under what scenarios would this happen out of curiosity? Don't we just want `g_assert (signatures != NULL);`?

Reply:
I tried this, but got an test failure:
```console
ERROR: tests/test-signed-pull.sh - Bail out! OSTreeSign:ERROR:src/libostree/ostree-sign-x509.c:169:ostree_sign_x509_data_verify: assertion failed: (signatures != NULL)
```
Looks like `ostree_sign_commit_verify` may call `ostree_sign_data_verify` with NULL `signatures`. Is there any case where this is supposed to succeed?

Reply:
Probably not but yeah let's continue to throw an error